[generate_vmap_rule] Add mechanism to override ctx.saved_tensors (CtxWithSavedTensors)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91211
* #91206
* #90966
* #90965
* __->__ #90964
* #90963

As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit#heading=h.r3ckcnsh1cxt

This PR creates CtxWithSavedTensors. You can wrap a ctx object in the
backward pass of autograd.Function in CtxWithSavedTensors and specify
the saved_tensors attribute. CtxWithSavedTensor acts like the original
ctx object (all other attribute accesses are forwarded to the original ctx
object) but it has a custom saved_tensors field.

Test Plan:
- tests that you can use CtxWithSavedTensors to get a new object with
your own saved_tensors.

zou3519 committed:
[generate_vmap_rule] Add mechanism to override ctx.saved_tensors (CtxWithSavedTensors)

As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit#heading=h.r3ckcnsh1cxt

This PR creates CtxWithSavedTensors. You can wrap a ctx object in the
backward pass of autograd.Function in CtxWithSavedTensors and specify
the saved_tensors attribute. CtxWithSavedTensor acts like the original
ctx object (all other attribute accesses are forwarded to the original ctx
object) but it has a custom saved_tensors field.

Test Plan:
- tests that you can use CtxWithSavedTensors to get a new object with
your own saved_tensors.

[ghstack-poisoned]

zou3519 committed:
Update on "[generate_vmap_rule] Add mechanism to override ctx.saved_tensors (CtxWithSavedTensors)"


As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit#heading=h.r3ckcnsh1cxt

This PR creates CtxWithSavedTensors. You can wrap a ctx object in the
backward pass of autograd.Function in CtxWithSavedTensors and specify
the saved_tensors attribute. CtxWithSavedTensor acts like the original
ctx object (all other attribute accesses are forwarded to the original ctx
object) but it has a custom saved_tensors field.

Test Plan:
- tests that you can use CtxWithSavedTensors to get a new object with
your own saved_tensors.

[ghstack-poisoned]

soulitzer commented:
See my other comment about possibly combining it with the other ctx wrapper since there is some shared code, but otherwise looks good.

zou3519 committed:
Update on "[generate_vmap_rule] Add mechanism to override ctx.saved_tensors (CtxWithSavedTensors)"


As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit#heading=h.r3ckcnsh1cxt

This PR creates CtxWithSavedTensors. You can wrap a ctx object in the
backward pass of autograd.Function in CtxWithSavedTensors and specify
the saved_tensors attribute. CtxWithSavedTensor acts like the original
ctx object (all other attribute accesses are forwarded to the original ctx
object) but it has a custom saved_tensors field.

Test Plan:
- tests that you can use CtxWithSavedTensors to get a new object with
your own saved_tensors.

[ghstack-poisoned]

zou3519 committed:
Update on "[generate_vmap_rule] Add mechanism to override ctx.saved_tensors (CtxWithSavedTensors)"


As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit#heading=h.r3ckcnsh1cxt

This PR creates CtxWithSavedTensors. You can wrap a ctx object in the
backward pass of autograd.Function in CtxWithSavedTensors and specify
the saved_tensors attribute. CtxWithSavedTensor acts like the original
ctx object (all other attribute accesses are forwarded to the original ctx
object) but it has a custom saved_tensors field.

Test Plan:
- tests that you can use CtxWithSavedTensors to get a new object with
your own saved_tensors.

[ghstack-poisoned]

zou3519 committed:
Update on "[generate_vmap_rule] Add mechanism to override ctx.saved_tensors (CtxWithSavedTensors)"


As seen in
https://docs.google.com/document/d/1bIQkWXy3J35_20c_a5kchikabBW5M8_uRAhl0BIMwU4/edit#heading=h.r3ckcnsh1cxt

This PR creates CtxWithSavedTensors. You can wrap a ctx object in the
backward pass of autograd.Function in CtxWithSavedTensors and specify
the saved_tensors attribute. CtxWithSavedTensor acts like the original
ctx object (all other attribute accesses are forwarded to the original ctx
object) but it has a custom saved_tensors field.

Test Plan:
- tests that you can use CtxWithSavedTensors to get a new object with
your own saved_tensors.

[ghstack-poisoned]
